filters/auth/grantlogout.go

```diff
@@ -0,0 +1,202 @@
+	if err == nil && buf != nil {
+		var errorResponse revokeErrorResponse
+		err = json.Unmarshal(buf, &errorResponse)
+
+		if err == nil {
+			return &errorResponse
+		}
+	}
+
+	return nil
+}
+
+func responseToError(response *http.Response, tokenType string) error {
+	errorResponse := getErrorResponse(response)
+	if errorResponse.Error == unsupportedTokenTypeError && tokenType == accessTokenType {
+		// Provider does not support revoking access tokens, which can happen according to RFC 7009.
+		// In that case this is not really an error.
+		return nil
+	}
+	return fmt.Errorf(
+		"%s revocation failed: %d %s",
+		tokenType,
+		response.StatusCode,
+		errorResponse.Error+": "+errorResponse.ErrorDescription,
+	)
+}
+
+func (f *grantLogoutFilter) revokeTokenType(tokenType string, token string) error {
+	revokeURL, err := url.Parse(f.config.RevokeTokenURL)
+	if err != nil {
+		return err
+	}
+
+	query := revokeURL.Query()
+	for k, v := range f.config.AuthURLParameters {
+		query.Set(k, v)
+	}
+	revokeURL.RawQuery = query.Encode()
+
+	body := url.Values{}
+	body.Add(revokeTokenKey, token)
+	body.Add(revokeTokenTypeKey, tokenType)
+
+	revokeRequest, err := http.NewRequest(
+		"POST",
+		revokeURL.String(),
+		strings.NewReader(body.Encode()))
+
+	if err != nil {
+		return err
+	}
+
+	authCredentials, err := f.getBasicAuthCredentials()
+	if err != nil {
+		return err
+	}
+
+	revokeRequest.SetBasicAuth(authCredentials.clientId, authCredentials.clientSecret)
+	revokeRequest.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	revokeResponse, err := f.config.AuthClient.Do(revokeRequest)
+	if err != nil {
+		return err
+	}
+	defer revokeResponse.Body.Close()
+
+	if revokeResponse.StatusCode == 400 {
+		return responseToError(revokeResponse, tokenType)
+	} else if revokeResponse.StatusCode != 200 {
+		return fmt.Errorf(
+			"%s revocation failed: %d",
+			tokenType,
+			revokeResponse.StatusCode,
+		)
+	}
+
+	return nil
+}
+
+func (f *grantLogoutFilter) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	c, err := extractCookie(req, f.config)
+	if err != nil {
+		unauthorized(
+			ctx,
+			"",
+			missingToken,
+			req.Host,
+			fmt.Sprintf("No token cookie %v in request.", f.config.TokenCookieName))
+		return
+	}
+
+	if c.AccessToken == "" && c.RefreshToken == "" {
+		unauthorized(
+			ctx,
+			"",
+			missingToken,
+			req.Host,
+			fmt.Sprintf("Token cookie %v has no tokens.", f.config.TokenCookieName))
+		return
+	}
+
+	var accessTokenRevokeError, refreshTokenRevokeError error
+	if c.AccessToken != "" {
+		accessTokenRevokeError = f.revokeTokenType(accessTokenType, c.AccessToken)
+		if accessTokenRevokeError != nil {
+			log.Error(err)
+		}
+	}
+
+	if c.RefreshToken != "" {
+		refreshTokenRevokeError = f.revokeTokenType(refreshTokenType, c.RefreshToken)
+		if refreshTokenRevokeError != nil {
+			log.Error(err)
+		}
+	}
+
+	if refreshTokenRevokeError != nil || accessTokenRevokeError != nil {
```

Comment:
From a human user perspective, I started thinking about whether we should return the 'ugly' proxy defined error page for every case. Maybe in some of the error cases it is fine to forward the request to the backend and let the user see a nice page from the application. E.g: if we cannot revoke a token because it's already expired. This also depends on how often we expect such cases, because if it happens too rarely, then probably not worth it. WDYT?

Reply:
I'm thinking that this should happen only rarely. 
If we tried to revoke a token that was expired already, this part of the RFC would actually come into play:
> Note: invalid tokens do not cause an error response since the client
   cannot handle such an error in a reasonable way.  Moreover, the
   purpose of the revocation request, invalidating the particular token,
   is already achieved.


The spec also considers this option:
> If the server responds with HTTP status code 503, the client must
   assume the token still exists and may retry after a reasonable delay.
   The server may include a "Retry-After" header in the response to
   indicate how long the service is expected to be unavailable to the
   requesting client.

This is something that we don't do, hence this would trigger the ugly error on our side. 

So the options are
- 200 (token revoked or the token was invalid to begin with)
- 400 (sever could not revoke the token) 
  - this might not cause an error on our side though if the server replied that it cannot revoke an auth_token. In that case, we treat this like a 200
- 503 (temporary overload), not handled gracefully on our side

So all in all my understanding is that this part of the code would only be run if the token response endpoint was behaving unexpectedly. Or of course if something in our implementation did not match the spec which would cause the requests to fail but we should be able to fix this on our side then.

Reply:
understood, thanks for the explanation!